azure-iot-device/tests/provisioning/pipeline/test_mqtt_topic_provisioning.py

```diff
@@ -0,0 +1,257 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+import pytest
+import logging
+from azure.iot.device.provisioning.pipeline import mqtt_topic_provisioning
+
+logging.basicConfig(level=logging.DEBUG)
+
+# NOTE: All tests (that require it) are parametrized with multiple values for URL encoding.
+# This is to show that the URL encoding is done correctly - not all URL encoding encodes
+# the same way. We must always test the ' ' and '/' characters specifically, in addition
+# to a generic URL encoding value (e.g. $, #, etc.)
```

Comment:
Do we not test decode (or encode) of + in this file?  The other big test file has a comment in the header about + but it's missing in this file.

It feels like we should just always test + for encode and decode so future developers don't need to think about "do I need to test + here or not".  If we just have every encode/decode test run against some list (maybe ' ', '$', '#', '/', '+', and, just for yucks, '\')

Reply:
No, counterintuitively we don't NEED to test for '+' for the following reason:

A '+' will always encode into %2B, and %2B will always decode to '+' no matter what style of URL encoding you do. 

HOWEVER, if using the plus style encoding, a ' ' will encode into a '+', and a '+' will decode back to a ' '.
In regular style encoding, a ' ' will encode into '%20', and then '%20' will decode back to a ' '

Furthermore, the extra consideration in Python is around the '/' due to the way that `quote_plus` by default considers '/' a "safety" character that is ignored. Wheras the standard `quote` does not, and so to ensure we are actually encoding as we desire, we must test it.

---
In reply to: [399808971](https://github.com/Azure/azure-iot-sdk-python/pull/514#discussion_r399808971) [](ancestors = 399808971)